File: src/main/java/Backtracking/Combinations.java
package main.java.Backtracking;

import java.util.List;
import java.util.ArrayList;

class Combinations {

    public static List<List<Integer>> generateCombinations(
            int[] nums,
            int r
    ) {
        List<List<Integer>> result = new ArrayList<>();

        backtracking(
                result,
                new ArrayList<>(),
                nums,
                0,
                r
        );

        return result;
    }

    public static List<List<Integer>> generateCombinations(
            int n,
            int r
    ) {
        int[] nums = new int[n];

        for (int i = 1; i <= n; i++) {
            nums[i - 1] = i;
        }

        return generateCombinations(nums, r);
    }

    private static void backtracking(
            List<List<Integer>> result,
            List<Integer> state,
            int[] nums,
            int idx,
            int r
    ) {
        // Successfully selected r elements
        if (r == 0) {
            result.add(new ArrayList<>(state));
            return;
        }

        // Ran out of elements before selecting r
        if (idx >= nums.length) {
            return;
        }

        for (int i = idx; i < nums.length; i++) {
            state.add(nums[i]);

            backtracking(
                    result,
                    state,
                    nums,
                    i + 1,
                    r - 1
            );

            state.remove(state.size() - 1);
        }
    }
}

File: src/main/python/Backtracking/combinations.py
class Combinations:

    @staticmethod
    def generate_combinations(
        nums: list[int],
        r: int
    ) -> list[list[int]]:

        result: list[list[int]] = []

        def __backtracking(
            state: list[int],
            idx: int,
            r: int
        ):
            # Successfully selected r elements
            if r == 0:
                result.append(state.copy())
                return

            # Ran out of elements before selecting r
            if idx >= len(nums):
                return

            for i in range(idx, len(nums)):
                state.append(nums[i])

                __backtracking(
                    state,
                    i + 1,
                    r - 1
                )

                state.pop()

        __backtracking([], 0, r)

        return result

    @staticmethod
    def generate_combinations_n(
        n: int,
        r: int
    ) -> list[list[int]]:

        nums: list[int] = []

        for i in range(1, n + 1):
            nums.append(i)

        return Combinations.generate_combinations(nums, r)